python/ray/data/preprocessor.py

```diff
@@ -218,6 +218,15 @@ def _determine_transform_to_use(self) -> BatchFormat:
             )
 
     def _transform(self, ds: "Dataset") -> "Dataset":
+        # Avoid direct import to avoid circular import
+        if ds.__class__.__name__ == "StreamSplitDataIterator":
+            raise ValueError(
+                "Cannot transform a StreamSplitDataIterator. This error typically occurs "
+                "when trying to apply a preprocessor to a dataset shard inside a train worker. "
+                "Instead, you should apply the transform to the Ray Dataset before sharding "
+                "it across workers."
+            )
```

Comment:
IMO might be more generalizable to do this directly at the `DataIterator` level, if any of the non-supported methods are called.